### PR TITLE
Legalize several i8 insts

### LINKED ISSUE #367: Please support load.i8 and store.i8 on x86

### PR DESCRIPTION
**Pull request by bjorn3**
Based on #379

Fixes #367

Implements for i8 and i16:

* iconst
* load
* store
* iadd
* isub
* imul
* udiv
* sdiv
* iadd_imm
* isub_imm
* imul_imm
* br_table

### CODE REVIEW

**Comment by bjorn3** [comment]
This allows me to remove all workarounds for missing i8 and i16 support.

File: lib/codegen/meta/base/legalize.py [hunk 1]
```diff
@@ -148,6 +156,86 @@
             a << iconcat(al, ah)
         ))
 
+for ty in [types.i8, types.i16]:
+    widen.legalize(
+        a << iconst.bind(ty)(b),
+        Rtl(
+            c << iconst.bind(types.i32)(b),
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply lib/codegen/meta/base/legalize.py]: Didn't know that

File: filetests/isa/x86/legalize-load-store-i8.cton [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    v3 = stack_addr.i64 ss1
+    store v1, v3
+    v4 = stack_addr.i64 ss2
+    store v2, v4
+    v5 = stack_addr.i64 ss3
+    v6 = stack_addr.i64 ss4
+    jump ebb1
+
+ebb1:
+; StorageLive(_3)
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply filetests/isa/x86/legalize-load-store-i8.cton]: Sure

File: lib/codegen/meta/base/legalize.py [hunk 3]
```diff
@@ -148,6 +156,86 @@
...
+    for ty in [types.i8, types.i16]:
+        widen.legalize(
+            a << binop.bind(ty)(x, y),
+            Rtl(
+                b << sextend.bind(types.i32)(x),
+                c << sextend.bind(types.i32)(y),
+                d << binop(b, c),
+                a << ireduce(d)
+            )
+        )
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply lib/codegen/meta/base/legalize.py]: Done

**Comment by bjorn3** [comment]
Why did it fail:

```
base/legalize.py:378: error: Incompatible types in assignment (expression has type "FloatType", variable has type "IntType")
```

I didn't change that line

**Comment by sunfishcode** [comment]
That's a mypy error, and I'm don't immediately know, but my guess is that it's inferring the type of `ty` to be `IntType` rather than `ValueType` or so, because the first appearance of `ty` is now iterating over integer types. That said, I don't yet know how to fix it.

**Comment by sunfishcode** [comment]
It might be enough to just namre the variable `int_ty` instead of `ty` when iterating over i8 and i16.

**Comment by sunfishcode** [comment]
The i8/i16 parts look good! This is now just waiting on the parts discussed in #379.

**Comment by sunfishcode** [comment]
Coincidentally, at this moment Cretonne is in the middle of a minor license adjustment process:

https://github.com/cretonne/cretonne/pull/361

Would you be willing to give permission to relicense your contributions?

If you agree, please reply and say "I agree to allow my contributions to the Cretonne project to be distributed under the terms of the "Apache-2.0 WITH LLVM-exception" license."

I'd be happy to answer any questions you may have about this process.

Thanks!

**Comment by bjorn3** [comment]
Rebased #378 away.

I agree to allow my contributions to the Cretonne project to be distributed under the terms of the "Apache-2.0 WITH LLVM-exception" license.

**Comment by sunfishcode** [comment]
Great! This PR looks good to go, except a few bits of the "nop" change still remain, causing test failures:

> FAIL filetests/isa/x86/legalize-bnot.cton: 21: Unknown opcode: 'nop'
FAIL filetests/isa/x86/legalize-imul-imm-i8.cton: 15: Unknown opcode: 'nop'
FAIL filetests/isa/x86/legalize-load-store-i8.cton: 22: Unknown opcode: 'nop'
FAIL filetests/isa/x86/legalize-regmove-i8.cton: 25: Unknown opcode: 'nop'

**Comment by bjorn3** [comment]
Fixed tests

**Comment by sunfishcode** [comment]
Thanks!

**Pull request merged**